In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
df_opc = pd.read_csv(r'C:\Users\boudl\Instacart Basket Analysis\02 Data\Prepared Data\orders_products_merged-0.csv')
#import dataframe

## Data Management

In [7]:
df_opc.shape

(32434212, 14)

Total dataframe with 32,434,212 pieces, chunked to four seperate dataframes for study for memory conservation.

In [14]:
df1 = pd.read_csv(r'C:\Users\boudl\Instacart Basket Analysis\02 Data\Prepared Data\orders_products_merged-0.csv', skiprows=0, nrows=8108553)
#chunk one for study

In [3]:
cols = pd.read_csv(r'C:\Users\boudl\Instacart Basket Analysis\02 Data\Prepared Data\orders_products_merged-0.csv', nrows=0).columns
#command to read first column header

In [22]:
df2 = pd.read_csv(r'C:\Users\boudl\Instacart Basket Analysis\02 Data\Prepared Data\orders_products_merged-0.csv', skiprows=range (1, 8108553), nrows=8108553)

In [11]:
df3 = pd.read_csv(r'C:\Users\boudl\Instacart Basket Analysis\02 Data\Prepared Data\orders_products_merged-0.csv', skiprows=16217106, nrows=8108553, header=None)
#chunk three for study

In [4]:
df4 = pd.read_csv(r'C:\Users\boudl\Instacart Basket Analysis\02 Data\Prepared Data\orders_products_merged-0.csv', skiprows=range (1, 24325659), nrows=8108553)
#chunk four for study

# df1 Study *********************************************************************

In [19]:
df1.head()
#command to study first five rows of dataframe

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   2539329        1             1            2                  8   
1   2398795        1             2            3                  7   
2    473747        1             3            3                 12   
3   2254736        1             4            4                  7   
4    431534        1             5            4                 15   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     NaN              True       196.0                1.0   
1                    15.0             False       196.0                1.0   
2                    21.0             False       196.0                1.0   
3                    29.0             False       196.0                1.0   
4                    28.0             False       196.0                1.0   

   reordered product_name  aisle_id  department_id  prices  
0        0.0         Soda        77              7     9.0  
1        1.0         Soda        77              7     9.0  
2        1.0         Soda        77              7     9.0  
3        1.0         Soda        77              7     9.0  
4        1.0         Soda        77              7     9.0

In [20]:
df1.groupby('department_id').agg({'order_number': ['mean']})
#group department id by mean number of orders

order_number
                      mean
department_id             
1                16.048520
3                18.487603
4                18.313599
5                16.339367
6                17.510720
7                16.926922
8                14.580000
9                14.497685
11               23.033037
12               17.276599
13               17.354712
14               16.736878
15               18.123020
16               18.326544
17               17.283393
18               19.784516
19               17.675713
20               17.175798

In [21]:
df1.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})
#group department id by mean min and max of order numbers

order_number        
                      mean min max
department_id                     
1                16.048520   1  99
3                18.487603   1  99
4                18.313599   1  99
5                16.339367   1  97
6                17.510720   1  99
7                16.926922   1  99
8                14.580000   1  80
9                14.497685   1  99
11               23.033037   1  99
12               17.276599   1  99
13               17.354712   1  99
14               16.736878   1  99
15               18.123020   1  99
16               18.326544   1  99
17               17.283393   1  99
18               19.784516   1  99
19               17.675713   1  99
20               17.175798   1  99

In [22]:
df1['max_order'] = df1.groupby(['user_id'])['order_number'].transform(np.max)
#transform data to create a customer loyalty column

C:\Users\boudl\AppData\Local\Temp\ipykernel_5408\1452403922.py:1: FutureWarning: The provided callable <function amax at 0x000001DC4B781C60> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df1['max_order'] = df1.groupby(['user_id'])['order_number'].transform(np.max)


Error occured using CF command string: The provided callable <function amax at 0x000001DC4B781C60> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df1['max_order'] = df1.groupby(['user_id'])['order_number'].transform(np.max)
  
This error indicates that future versions of pandas may not use or recognize this command string. Thus, a different command will be used henceforth

In [23]:
df1['max_order'] = df1.groupby(['user_id'])['order_number'].transform('max')
#transfrom data to create a customer loyalty column

In [24]:
df1.head()
#command to study changes

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   2539329        1             1            2                  8   
1   2398795        1             2            3                  7   
2    473747        1             3            3                 12   
3   2254736        1             4            4                  7   
4    431534        1             5            4                 15   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     NaN              True       196.0                1.0   
1                    15.0             False       196.0                1.0   
2                    21.0             False       196.0                1.0   
3                    29.0             False       196.0                1.0   
4                    28.0             False       196.0                1.0   

   reordered product_name  aisle_id  department_id  prices  max_order  
0        0.0         Soda        77              7     9.0         10  
1        1.0         Soda        77              7     9.0         10  
2        1.0         Soda        77              7     9.0         10  
3        1.0         Soda        77              7     9.0         10  
4        1.0         Soda        77              7     9.0         10

In [25]:
df1.loc[df1['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

C:\Users\boudl\AppData\Local\Temp\ipykernel_5408\3559291325.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'Loyal customer' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df1.loc[df1['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'


CF command caused this error: Value 'Loyal customer' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df1.loc[df1['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer
  
 Proceeded to change dtype then continue with commands to create flag column

In [28]:
df1['loyalty_flag'] = df1.get('loyalty_flag', pd.Series(dtype='object'))
#change column dtype to conduct creation of flag column

In [29]:
df1.loc[df1['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
#assigning first rule of flag column

In [30]:
df1.loc[(df1['max_order'] <= 40) & (df1['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
#assigning second rule of flag column

In [31]:
df1.loc[df1['max_order'] <= 10, 'loyalty_flag'] = 'New customer'
#assigning tihrd rule of flag column

In [32]:
df1['loyalty_flag'].value_counts(dropna = False)
#command to display counts of flag column with applied rules

loyalty_flag
Regular customer    3898409
Loyal customer      2765085
New customer        1445059
Name: count, dtype: int64

In [34]:
df1.head(15)
#confirm changes in loyalty flag column

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0    2539329        1             1            2                  8   
1    2398795        1             2            3                  7   
2     473747        1             3            3                 12   
3    2254736        1             4            4                  7   
4     431534        1             5            4                 15   
5    3367565        1             6            2                  7   
6     550135        1             7            1                  9   
7    3108588        1             8            1                 14   
8    2295261        1             9            1                 16   
9    2550362        1            10            4                  8   
10   2968173       15            15            1                  9   
11   1870022       15            17            2                 16   
12   1911383       15            18            2                 11   
13   2715276       15            21            1                  9   
14    487368       15            22            1                 10   

    days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                      NaN              True       196.0                1.0   
1                     15.0             False       196.0                1.0   
2                     21.0             False       196.0                1.0   
3                     29.0             False       196.0                1.0   
4                     28.0             False       196.0                1.0   
5                     19.0             False       196.0                1.0   
6                     20.0             False       196.0                1.0   
7                     14.0             False       196.0                2.0   
8                      0.0             False       196.0                4.0   
9                     30.0             False       196.0                1.0   
10                     7.0             False       196.0                2.0   
11                     8.0             False       196.0                6.0   
12                     7.0             False       196.0                1.0   
13                     7.0             False       196.0                1.0   
14                    14.0             False       196.0                1.0   

    reordered product_name  aisle_id  department_id  prices  max_order  \
0         0.0         Soda        77              7     9.0         10   
1         1.0         Soda        77              7     9.0         10   
2         1.0         Soda        77              7     9.0         10   
3         1.0         Soda        77              7     9.0         10   
4         1.0         Soda        77              7     9.0         10   
5         1.0         Soda        77              7     9.0         10   
6         1.0         Soda        77              7     9.0         10   
7         1.0         Soda        77              7     9.0         10   
8         1.0         Soda        77              7     9.0         10   
9         1.0         Soda        77              7     9.0         10   
10        0.0         Soda        77              7     9.0         22   
11        1.0         Soda        77              7     9.0         22   
12        1.0         Soda        77              7     9.0         22   
13        1.0         Soda        77              7     9.0         22   
14        1.0         Soda        77              7     9.0         22   

        loyalty_flag  
0       New customer  
1       New customer  
2       New customer  
3       New customer  
4       New customer  
5       New customer  
6       New customer  
7       New customer  
8       New customer  
9       New customer  
10  Regular customer  
11  Regular customer  
12  Regular customer  
13  Regular customer  
14  Regular customer

In [37]:
spending_loyalty = df1.groupby('loyalty_flag')['prices'].agg(['mean', 'median', 'std'])
#command to study spending habits of customers

In [38]:
spending_loyalty.head()
#command to study mean median and standard deviation of customer spending habits

mean  median       std
loyalty_flag                                
Loyal customer    8.158956     8.2  4.203864
New customer      8.269837     8.3  4.281930
Regular customer  8.247409     8.2  4.268591

In [40]:
df1['avg_price_per_user'] = df1.groupby('user_id')['prices'].transform('mean')
#calc average price of orders by customer

In [44]:
df1['spending_flag'] = df1.get('spending_flag', pd.Series(dtype='object'))
#command to change dtype of column before using transform function

In [45]:
df1.loc[df1['avg_price_per_user'] < 10, 'spending_flag'] = 'Low spender'
#first rule of flag column

In [46]:
df1.loc[df1['avg_price_per_user'] >= 10, 'spending_flag'] = 'High spender'
#second rule of flag column

In [51]:
df1['spending_flag'].value_counts(dropna = False)
#command to study counts of new flag column

spending_flag
Low spender     7540714
High spender     567839
Name: count, dtype: int64

In [49]:
df1.head(10)
#command to study dataframe shape with new flag column

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   2539329        1             1            2                  8   
1   2398795        1             2            3                  7   
2    473747        1             3            3                 12   
3   2254736        1             4            4                  7   
4    431534        1             5            4                 15   
5   3367565        1             6            2                  7   
6    550135        1             7            1                  9   
7   3108588        1             8            1                 14   
8   2295261        1             9            1                 16   
9   2550362        1            10            4                  8   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     NaN              True       196.0                1.0   
1                    15.0             False       196.0                1.0   
2                    21.0             False       196.0                1.0   
3                    29.0             False       196.0                1.0   
4                    28.0             False       196.0                1.0   
5                    19.0             False       196.0                1.0   
6                    20.0             False       196.0                1.0   
7                    14.0             False       196.0                2.0   
8                     0.0             False       196.0                4.0   
9                    30.0             False       196.0                1.0   

   reordered product_name  aisle_id  department_id  prices  max_order  \
0        0.0         Soda        77              7     9.0         10   
1        1.0         Soda        77              7     9.0         10   
2        1.0         Soda        77              7     9.0         10   
3        1.0         Soda        77              7     9.0         10   
4        1.0         Soda        77              7     9.0         10   
5        1.0         Soda        77              7     9.0         10   
6        1.0         Soda        77              7     9.0         10   
7        1.0         Soda        77              7     9.0         10   
8        1.0         Soda        77              7     9.0         10   
9        1.0         Soda        77              7     9.0         10   

   loyalty_flag  avg_price_per_user spending_flag  
0  New customer            6.367797   Low spender  
1  New customer            6.367797   Low spender  
2  New customer            6.367797   Low spender  
3  New customer            6.367797   Low spender  
4  New customer            6.367797   Low spender  
5  New customer            6.367797   Low spender  
6  New customer            6.367797   Low spender  
7  New customer            6.367797   Low spender  
8  New customer            6.367797   Low spender  
9  New customer            6.367797   Low spender

In [70]:
df1['user_frequency'] = df1.groupby('user_id')['days_since_prior_order'].transform('median')
#calc median value of user frequency 

In [71]:
df1['user_frequency'] = pd.Series(dtype='object')
#command to change dtype of column

In [72]:
df1.loc[df1['median_days_since_prior'] > 20, 'user_frequency'] = 'Non-frequent customer'
#first rule of flag column

In [73]:
df1.loc[(df1['median_days_since_prior'] > 10) & (df1['median_days_since_prior'] <= 20), 'user_frequency'] = 'Regular customer'
#second rule of flag column

In [74]:
df1.loc[df1['median_days_since_prior'] <= 10, 'user_frequency'] = 'Frequent customer'
#third rule of flag column

In [75]:
df1['user_frequency'].value_counts(dropna=False)
#command to display new flag column counts

user_frequency
Frequent customer        5568889
Regular customer         1700688
Non-frequent customer     834961
NaN                         4015
Name: count, dtype: int64

In [76]:
df1.head()
#command to display dataframe with new flag column

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   2539329        1             1            2                  8   
1   2398795        1             2            3                  7   
2    473747        1             3            3                 12   
3   2254736        1             4            4                  7   
4    431534        1             5            4                 15   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     NaN              True       196.0                1.0   
1                    15.0             False       196.0                1.0   
2                    21.0             False       196.0                1.0   
3                    29.0             False       196.0                1.0   
4                    28.0             False       196.0                1.0   

   reordered product_name  aisle_id  department_id  prices  max_order  \
0        0.0         Soda        77              7     9.0         10   
1        1.0         Soda        77              7     9.0         10   
2        1.0         Soda        77              7     9.0         10   
3        1.0         Soda        77              7     9.0         10   
4        1.0         Soda        77              7     9.0         10   

   loyalty_flag  avg_price_per_user spending_flag         user_frequency  \
0  New customer            6.367797   Low spender  Non-frequent customer   
1  New customer            6.367797   Low spender  Non-frequent customer   
2  New customer            6.367797   Low spender  Non-frequent customer   
3  New customer            6.367797   Low spender  Non-frequent customer   
4  New customer            6.367797   Low spender  Non-frequent customer   

   median_days_since_prior  
0                     20.5  
1                     20.5  
2                     20.5  
3                     20.5  
4                     20.5

In [77]:
path = r'C:\Users\boudl\Instacart Basket Analysis'

In [78]:
df1.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'OPCC' ,'opc_chunk1.csv'))

## df2 study ***************************************************************

In [23]:
df2.head(2)
#command to study first first five rows of dataframe

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   1155044   115423            62            3                 19   
1    967065   115423            68            2                 23   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     2.0             False     36070.0               12.0   
1                     4.0             False     36070.0               13.0   

   reordered                                       product_name  aisle_id  \
0        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
1        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   

   department_id  prices  
0              4     9.9  
1              4     9.9

In [24]:
df2.groupby('department_id').agg({'order_number': ['mean']})
#group department id by mean number of orders

order_number
                      mean
department_id             
1                15.838305
2                17.604733
3                17.296235
4                17.540484
5                15.595915
6                16.197823
7                17.536728
8                17.550776
9                16.269315
11               16.342697
12               15.711600
13               16.442890
14               16.794095
15               16.095261
16               17.622183
17               15.662408
18               19.111308
19               17.400434
20               16.167257
21               23.576042

In [25]:
df2.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})
#group department id by mean min and max of order numbers

order_number        
                      mean min max
department_id                     
1                15.838305   1  99
2                17.604733   1  99
3                17.296235   1  99
4                17.540484   1  99
5                15.595915   1  99
6                16.197823   1  99
7                17.536728   1  99
8                17.550776   1  97
9                16.269315   1  99
11               16.342697   1  99
12               15.711600   1  99
13               16.442890   1  99
14               16.794095   1  99
15               16.095261   1  99
16               17.622183   1  99
17               15.662408   1  99
18               19.111308   1  99
19               17.400434   1  99
20               16.167257   1  99
21               23.576042   1  99

In [26]:
df2['max_order'] = df2.groupby(['user_id'])['order_number'].transform('max')
#transfrom data to create a customer loyalty column

In [27]:
df2.head()
#command to study changes

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   1155044   115423            62            3                 19   
1    967065   115423            68            2                 23   
2   2109811   115423            70            3                 15   
3   1723180   115423            78            4                 23   
4   1131277   115423            86            3                  9   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     2.0             False     36070.0               12.0   
1                     4.0             False     36070.0               13.0   
2                     6.0             False     36070.0               27.0   
3                     0.0             False     36070.0                5.0   
4                     1.0             False     36070.0                9.0   

   reordered                                       product_name  aisle_id  \
0        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
1        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
2        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
3        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
4        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   

   department_id  prices  max_order  
0              4     9.9         99  
1              4     9.9         99  
2              4     9.9         99  
3              4     9.9         99  
4              4     9.9         99

In [28]:
df2['loyalty_flag'] = df2.get('loyalty_flag', pd.Series(dtype='object'))
#change column dtype to conduct creation of flag column

In [29]:
df2.loc[df2['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
#assigning first rule of flag column

In [31]:
df2.loc[(df2['max_order'] <= 40) & (df2['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
#assigning second rule of flag column

In [32]:
df2.loc[df2['max_order'] <= 10, 'loyalty_flag'] = 'New customer'
#assigning tihrd rule of flag column

In [33]:
df2['loyalty_flag'].value_counts(dropna = False)
#command to display counts of flag column with applied rules

loyalty_flag
Regular customer    3946246
Loyal customer      2579529
New customer        1582778
Name: count, dtype: int64

In [34]:
df2.head(15)
#confirm changes in loyalty flag column

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0    1155044   115423            62            3                 19   
1     967065   115423            68            2                 23   
2    2109811   115423            70            3                 15   
3    1723180   115423            78            4                 23   
4    1131277   115423            86            3                  9   
5    3078073   115423            91            3                 19   
6    2699772   115423            95            4                 19   
7    1872348   115423            98            4                 14   
8    2685028   115495            47            6                 12   
9    1398059   115507            11            0                 15   
10   2042286   115514            12            4                 10   
11   1992490   115537             1            0                 15   
12   3334187   115537             3            0                 12   
13   3418540   115552            15            6                 15   
14   2055847   115553             7            2                 10   

    days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                      2.0             False     36070.0               12.0   
1                      4.0             False     36070.0               13.0   
2                      6.0             False     36070.0               27.0   
3                      0.0             False     36070.0                5.0   
4                      1.0             False     36070.0                9.0   
5                      0.0             False     36070.0                5.0   
6                      5.0             False     36070.0                6.0   
7                      1.0             False     36070.0                6.0   
8                     12.0             False     36070.0               39.0   
9                      7.0             False     36070.0               10.0   
10                    30.0             False     36070.0               19.0   
11                     NaN              True     36070.0               20.0   
12                     7.0             False     36070.0               12.0   
13                     6.0             False     36070.0                8.0   
14                     1.0             False     36070.0                7.0   

    reordered                                       product_name  aisle_id  \
0         1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
1         1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
2         1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
3         1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
4         1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
5         1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
6         1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
7         1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
8         0.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
9         0.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
10        0.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
11        0.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
12        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
13        0.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
14        0.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   

    department_id  prices  max_order      loyalty_flag  
0               4     9.9         99    Loyal customer  
1               4     9.9         99    Loyal customer  
2               4     9.9         99    Loyal customer  
3               4     9.9         99    Loyal customer  
4               4     9.9         99    Loyal customer  
5          

In [35]:
spending_loyalty = df2.groupby('loyalty_flag')['prices'].agg(['mean', 'median', 'std'])
#command to study spending habits of customers

In [36]:
spending_loyalty.head()
#command to study mean median and standard deviation of customer spending habits

mean  median         std
loyalty_flag                                   
Loyal customer    14.869646     7.2  326.725739
New customer      16.641792     7.2  365.914317
Regular customer  16.074861     7.2  354.020433

In [37]:
df2['avg_price_per_user'] = df2.groupby('user_id')['prices'].transform('mean')
#calc average price of orders by customer

In [38]:
df2['spending_flag'] = df2.get('spending_flag', pd.Series(dtype='object'))
#command to change dtype of column before using transform function

In [39]:
df2.loc[df2['avg_price_per_user'] < 10, 'spending_flag'] = 'Low spender'
#first rule of flag column

In [40]:
df2.loc[df2['avg_price_per_user'] >= 10, 'spending_flag'] = 'High spender'
#second rule of flag column

In [41]:
df2['spending_flag'].value_counts(dropna = False)
#command to study counts of new flag column

spending_flag
Low spender     7701385
High spender     407168
Name: count, dtype: int64

In [42]:
df2.head(10)
#command to study dataframe shape with new flag column

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   1155044   115423            62            3                 19   
1    967065   115423            68            2                 23   
2   2109811   115423            70            3                 15   
3   1723180   115423            78            4                 23   
4   1131277   115423            86            3                  9   
5   3078073   115423            91            3                 19   
6   2699772   115423            95            4                 19   
7   1872348   115423            98            4                 14   
8   2685028   115495            47            6                 12   
9   1398059   115507            11            0                 15   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     2.0             False     36070.0               12.0   
1                     4.0             False     36070.0               13.0   
2                     6.0             False     36070.0               27.0   
3                     0.0             False     36070.0                5.0   
4                     1.0             False     36070.0                9.0   
5                     0.0             False     36070.0                5.0   
6                     5.0             False     36070.0                6.0   
7                     1.0             False     36070.0                6.0   
8                    12.0             False     36070.0               39.0   
9                     7.0             False     36070.0               10.0   

   reordered                                       product_name  aisle_id  \
0        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
1        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
2        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
3        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
4        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
5        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
6        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
7        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
8        0.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
9        0.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   

   department_id  prices  max_order      loyalty_flag  avg_price_per_user  \
0              4     9.9         99    Loyal customer            6.925091   
1              4     9.9         99    Loyal customer            6.925091   
2              4     9.9         99    Loyal customer            6.925091   
3              4     9.9         99    Loyal customer            6.925091   
4              4     9.9         99    Loyal customer            6.925091   
5              4     9.9         99    Loyal customer            6.925091   
6              4     9.9         99    Loyal customer            6.925091   
7              4     9.9         99    Loyal customer            6.925091   
8              4     9.9         57    Loyal customer            8.045014   
9              4     9.9         32  Regular customer            8.343902   

  spending_flag  
0   Low spender  
1   Low spender  
2   Low spender  
3   Low spender  
4   Low spender  
5   Low spender  
6   Low spender  
7   Low spender  
8   Low spender  
9   Low spender

In [55]:
# Calculate the median of 'days_since_prior_order' for each 'user_id' and store it in 'user_frequency'
df2['user_frequency'] = df2.groupby('user_id')['days_since_prior_order'].transform('median')

In [56]:
# Create a new column 'median_days_since_prior' based on the calculated median
df2['median_days_since_prior'] = df2['user_frequency']

In [58]:
df2.loc[df2['median_days_since_prior'] > 20, 'user_frequency'] = 'Non-frequent customer'
#first rule of flag column

In [59]:
df2.loc[(df2['median_days_since_prior'] > 10) & (df2['median_days_since_prior'] <= 20), 'user_frequency'] = 'Regular customer'
#second rule of flag column

In [61]:
df2.loc[df2['median_days_since_prior'] <= 10, 'user_frequency'] = 'Frequent customer'
#third rule of flag column

In [62]:
df2['user_frequency'].value_counts(dropna=False)
# Command to display new flag column counts

user_frequency
Frequent customer        5372690
Regular customer         1791673
Non-frequent customer     941400
NaN                         2790
Name: count, dtype: int64

In [63]:
df2.head()
#command to display new flag column

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   1155044   115423            62            3                 19   
1    967065   115423            68            2                 23   
2   2109811   115423            70            3                 15   
3   1723180   115423            78            4                 23   
4   1131277   115423            86            3                  9   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     2.0             False     36070.0               12.0   
1                     4.0             False     36070.0               13.0   
2                     6.0             False     36070.0               27.0   
3                     0.0             False     36070.0                5.0   
4                     1.0             False     36070.0                9.0   

   reordered                                       product_name  aisle_id  \
0        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
1        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
2        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
3        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   
4        1.0  Super Spinach! Baby Spinach, Baby Bok Choy, Sw...       123   

   department_id  prices  max_order    loyalty_flag  avg_price_per_user  \
0              4     9.9         99  Loyal customer            6.925091   
1              4     9.9         99  Loyal customer            6.925091   
2              4     9.9         99  Loyal customer            6.925091   
3              4     9.9         99  Loyal customer            6.925091   
4              4     9.9         99  Loyal customer            6.925091   

  spending_flag     user_frequency  median_days_since_prior  
0   Low spender  Frequent customer                      4.0  
1   Low spender  Frequent customer                      4.0  
2   Low spender  Frequent customer                      4.0  
3   Low spender  Frequent customer                      4.0  
4   Low spender  Frequent customer                      4.0

In [64]:
path = r'C:\Users\boudl\Instacart Basket Analysis'

In [65]:
df2.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'OPCC' ,'opc_chunk2.csv'))

## df3 Study ***********************************************************

In [12]:
df3.columns = cols
#assign headers to dataframe three

In [14]:
df3.head(2)
#command to study first two rows of dataframe

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   3033125   115439             6            2                 16   
1   3078491   115591             3            2                 15   
2   2441959   115619             1            6                  8   
3   2307024   115619             3            2                 16   
4   2945810   115619             7            3                 11   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     8.0             False     25544.0               11.0   
1                    11.0             False     25544.0                1.0   
2                     NaN              True     25544.0               20.0   
3                     5.0             False     25544.0               26.0   
4                    20.0             False     25544.0               30.0   

   reordered                              product_name  aisle_id  \
0        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
1        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
2        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
3        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
4        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   

   department_id  prices  
0             15     5.5  
1             15     5.5  
2             15     5.5  
3             15     5.5  
4             15     5.5

In [15]:
df3.groupby('department_id').agg({'order_number': ['mean']})
#group department id by mean number of orders

order_number
                      mean
department_id             
1                15.947210
2                19.615562
3                17.408827
4                16.771178
5                15.409562
6                16.489038
7                17.276593
8                14.884031
9                16.009594
10               20.803259
11               16.167277
12               16.254818
13               16.880087
14               17.056699
15               16.409371
16               18.045590
17               16.152323
18               19.953627
19               17.160756
20               15.871914
21               22.866904

In [16]:
df3.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})
#group department id by mean min and max of order numbers

order_number        
                      mean min max
department_id                     
1                15.947210   1  99
2                19.615562   1  99
3                17.408827   1  99
4                16.771178   1  99
5                15.409562   1  99
6                16.489038   1  99
7                17.276593   1  99
8                14.884031   1  99
9                16.009594   1  99
10               20.803259   1  99
11               16.167277   1  99
12               16.254818   1  99
13               16.880087   1  99
14               17.056699   1  99
15               16.409371   1  99
16               18.045590   1  99
17               16.152323   1  99
18               19.953627   1  99
19               17.160756   1  99
20               15.871914   1  99
21               22.866904   1  99

In [17]:
df3['max_order'] = df3.groupby(['user_id'])['order_number'].transform('max')
#transfrom data to create a customer loyalty column

In [19]:
df3.head()
#command to study changes

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   3033125   115439             6            2                 16   
1   3078491   115591             3            2                 15   
2   2441959   115619             1            6                  8   
3   2307024   115619             3            2                 16   
4   2945810   115619             7            3                 11   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     8.0             False     25544.0               11.0   
1                    11.0             False     25544.0                1.0   
2                     NaN              True     25544.0               20.0   
3                     5.0             False     25544.0               26.0   
4                    20.0             False     25544.0               30.0   

   reordered                              product_name  aisle_id  \
0        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
1        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
2        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
3        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
4        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   

   department_id  prices  max_order  
0             15     5.5          7  
1             15     5.5          3  
2             15     5.5         11  
3             15     5.5         11  
4             15     5.5         11

In [20]:
df3['loyalty_flag'] = df3.get('loyalty_flag', pd.Series(dtype='object'))
#change column dtype to conduct creation of flag column

In [21]:
df3.loc[df3['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
#assigning first rule of flag column

In [22]:
df3.loc[(df3['max_order'] <= 40) & (df3['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
#assigning second rule of flag column

In [23]:
df3.loc[df3['max_order'] <= 10, 'loyalty_flag'] = 'New customer'
#assigning tihrd rule of flag column

In [24]:
df3['loyalty_flag'].value_counts(dropna = False)
#command to display counts of flag column with applied rules

loyalty_flag
Regular customer    3947890
Loyal customer      2585807
New customer        1574856
Name: count, dtype: int64

In [25]:
df3.head(15)
#confirm changes in loyalty flag column

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0    3033125   115439             6            2                 16   
1    3078491   115591             3            2                 15   
2    2441959   115619             1            6                  8   
3    2307024   115619             3            2                 16   
4    2945810   115619             7            3                 11   
5    3374426   115633             5            6                 12   
6    1534082   115643             6            0                 10   
7    3199187   115659            27            1                 13   
8     305219   115659            42            0                 15   
9    2636121   115693             9            0                 16   
10    246330   115693            11            3                 19   
11   3299666   115693            13            5                  9   
12   2409361   115693            16            3                 18   
13   2922643   115693            17            4                  9   
14   2840818   115693            18            5                  8   

    days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                      8.0             False     25544.0               11.0   
1                     11.0             False     25544.0                1.0   
2                      NaN              True     25544.0               20.0   
3                      5.0             False     25544.0               26.0   
4                     20.0             False     25544.0               30.0   
5                     22.0             False     25544.0               10.0   
6                     30.0             False     25544.0                6.0   
7                      6.0             False     25544.0                5.0   
8                      6.0             False     25544.0                8.0   
9                      9.0             False     25544.0               22.0   
10                    11.0             False     25544.0               25.0   
11                     7.0             False     25544.0               18.0   
12                     7.0             False     25544.0               19.0   
13                     8.0             False     25544.0               29.0   
14                     8.0             False     25544.0               18.0   

    reordered                              product_name  aisle_id  \
0         0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
1         0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
2         0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
3         1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
4         1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
5         0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
6         0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
7         0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
8         1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
9         0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
10        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
11        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
12        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
13        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
14        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   

    department_id  prices  max_order      loyalty_flag  
0              15     5.5          7      New customer  
1              15     5.5          3      New customer  
2              15     5.5         11  Regular customer  
3              15     5.5         11  Regular customer  
4              15     5.5         11  Regular customer  
5              15     5.5          5      New customer  
6              15     5.5          7      New customer  
7              15     5.5         49    L

In [27]:
spending_loyalty = df3.groupby('loyalty_flag')['prices'].agg(['mean', 'median', 'std'])
#command to study spending habits of customers

In [28]:
spending_loyalty.head()
#command to study mean median and standard deviation of customer spending habits

mean  median       std
loyalty_flag                                
Loyal customer    7.692169     7.0  4.227659
New customer      7.705973     7.0  4.325741
Regular customer  7.716333     7.1  4.297337

In [29]:
df3['avg_price_per_user'] = df3.groupby('user_id')['prices'].transform('mean')
#calc average price of orders by customer

In [30]:
df3['spending_flag'] = df3.get('spending_flag', pd.Series(dtype='object'))
#command to change dtype of column before using transform function

In [32]:
df3.loc[df3['avg_price_per_user'] < 10, 'spending_flag'] = 'Low spender'
#first rule of flag column

In [33]:
df3.loc[df3['avg_price_per_user'] >= 10, 'spending_flag'] = 'High spender'
#second rule of flag column

In [34]:
df3['spending_flag'].value_counts(dropna = False)
#command to study counts of new flag column

spending_flag
Low spender     7755391
High spender     353162
Name: count, dtype: int64

In [35]:
df3.head(10)
#command to study dataframe shape with new flag column

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   3033125   115439             6            2                 16   
1   3078491   115591             3            2                 15   
2   2441959   115619             1            6                  8   
3   2307024   115619             3            2                 16   
4   2945810   115619             7            3                 11   
5   3374426   115633             5            6                 12   
6   1534082   115643             6            0                 10   
7   3199187   115659            27            1                 13   
8    305219   115659            42            0                 15   
9   2636121   115693             9            0                 16   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     8.0             False     25544.0               11.0   
1                    11.0             False     25544.0                1.0   
2                     NaN              True     25544.0               20.0   
3                     5.0             False     25544.0               26.0   
4                    20.0             False     25544.0               30.0   
5                    22.0             False     25544.0               10.0   
6                    30.0             False     25544.0                6.0   
7                     6.0             False     25544.0                5.0   
8                     6.0             False     25544.0                8.0   
9                     9.0             False     25544.0               22.0   

   reordered                              product_name  aisle_id  \
0        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
1        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
2        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
3        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
4        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
5        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
6        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
7        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
8        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
9        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   

   department_id  prices  max_order      loyalty_flag  avg_price_per_user  \
0             15     5.5          7      New customer            7.410345   
1             15     5.5          3      New customer            4.166667   
2             15     5.5         11  Regular customer            7.146914   
3             15     5.5         11  Regular customer            7.146914   
4             15     5.5         11  Regular customer            7.146914   
5             15     5.5          5      New customer            8.981250   
6             15     5.5          7      New customer            8.673077   
7             15     5.5         49    Loyal customer            7.226126   
8             15     5.5         49    Loyal customer            7.226126   
9             15     5.5         35  Regular customer            9.443154   

  spending_flag  
0   Low spender  
1   Low spender  
2   Low spender  
3   Low spender  
4   Low spender  
5   Low spender  
6   Low spender  
7   Low spender  
8   Low spender  
9   Low spender

In [36]:
df3['user_frequency'] = df3.groupby('user_id')['days_since_prior_order'].transform('median')
#calc median of days since prior order column by user id to create new column

In [37]:
df3['median_days_since_prior'] = df3['user_frequency']
#create new column based on user frequency column

In [38]:
df3.loc[df3['median_days_since_prior'] > 20, 'user_frequency'] = 'Non-frequent customer'
#first rule of flag column

C:\Users\boudl\AppData\Local\Temp\ipykernel_20528\2341751797.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'Non-frequent customer' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df3.loc[df3['median_days_since_prior'] > 20, 'user_frequency'] = 'Non-frequent customer'


In [39]:
df3.loc[(df3['median_days_since_prior'] > 10) & (df3['median_days_since_prior'] <= 20), 'user_frequency'] = 'Regular customer'
#second rule of flag column

In [40]:
df3.loc[df3['median_days_since_prior'] <= 10, 'user_frequency'] = 'Frequent customer'
#third rule of flag column

In [41]:
df3['user_frequency'].value_counts(dropna=False)
# Command to display new flag column counts

user_frequency
Frequent customer        5369471
Regular customer         1799103
Non-frequent customer     937459
NaN                         2520
Name: count, dtype: int64

In [42]:
df3.head()
#command to display new flag column

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   3033125   115439             6            2                 16   
1   3078491   115591             3            2                 15   
2   2441959   115619             1            6                  8   
3   2307024   115619             3            2                 16   
4   2945810   115619             7            3                 11   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     8.0             False     25544.0               11.0   
1                    11.0             False     25544.0                1.0   
2                     NaN              True     25544.0               20.0   
3                     5.0             False     25544.0               26.0   
4                    20.0             False     25544.0               30.0   

   reordered                              product_name  aisle_id  \
0        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
1        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
2        0.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
3        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   
4        1.0  Fresh Cut Golden Sweet Whole Kernel Corn        81   

   department_id  prices  max_order      loyalty_flag  avg_price_per_user  \
0             15     5.5          7      New customer            7.410345   
1             15     5.5          3      New customer            4.166667   
2             15     5.5         11  Regular customer            7.146914   
3             15     5.5         11  Regular customer            7.146914   
4             15     5.5         11  Regular customer            7.146914   

  spending_flag     user_frequency  median_days_since_prior  
0   Low spender   Regular customer                     17.0  
1   Low spender   Regular customer                     11.0  
2   Low spender  Frequent customer                      5.0  
3   Low spender  Frequent customer                      5.0  
4   Low spender  Frequent customer                      5.0

In [43]:
path = r'C:\Users\boudl\Instacart Basket Analysis'

In [44]:
df3.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'OPCC' ,'opc_chunk3.csv'))

## df4 Study *******************************************

In [5]:
df4.columns = cols
#assign headers to dataframe four

In [6]:
df4.head(2)
#command to study first two rows of dataframe

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   2059432    80013            11            1                 16   
1     40246    80029            18            0                  9   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                    14.0             False     26842.0                3.0   
1                    11.0             False     26842.0                6.0   

   reordered                 product_name  aisle_id  department_id  prices  
0        1.0  Boneless Beef Sirloin Steak       122             12    20.1  
1        0.0  Boneless Beef Sirloin Steak       122             12    20.1

In [7]:
df4.groupby('department_id').agg({'order_number': ['mean']})
#group department id by mean number of orders

order_number
                      mean
department_id             
1                14.673377
2                17.111714
3                16.755195
4                16.131260
5                14.980902
6                16.426027
7                16.881247
8                15.204871
9                15.503626
10               19.046609
11               16.145282
12               14.930996
13               16.142114
14               16.470926
15               15.237003
16               16.377953
17               15.192677
18               18.729026
19               16.958890
20               16.476734
21               22.742812

In [8]:
df4.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})
#group department id by mean min and max of order numbers

order_number        
                      mean min max
department_id                     
1                14.673377   1  99
2                17.111714   1  99
3                16.755195   1  99
4                16.131260   1  99
5                14.980902   1  99
6                16.426027   1  99
7                16.881247   1  99
8                15.204871   1  99
9                15.503626   1  99
10               19.046609   1  99
11               16.145282   1  99
12               14.930996   1  99
13               16.142114   1  99
14               16.470926   1  99
15               15.237003   1  99
16               16.377953   1  99
17               15.192677   1  99
18               18.729026   1  99
19               16.958890   1  99
20               16.476734   1  99
21               22.742812   1  99

In [9]:
df4['max_order'] = df4.groupby(['user_id'])['order_number'].transform('max')
#transfrom data to create a customer loyalty column

In [10]:
df4.head()
#command to study changes

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   2059432    80013            11            1                 16   
1     40246    80029            18            0                  9   
2    648191    80154             4            4                 17   
3   1509851    80263            13            5                  1   
4   1500945    80263            15            0                 11   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                    14.0             False     26842.0                3.0   
1                    11.0             False     26842.0                6.0   
2                    19.0             False     26842.0                7.0   
3                    20.0             False     26842.0                3.0   
4                     4.0             False     26842.0                2.0   

   reordered                 product_name  aisle_id  department_id  prices  \
0        1.0  Boneless Beef Sirloin Steak       122             12    20.1   
1        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
2        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
3        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
4        1.0  Boneless Beef Sirloin Steak       122             12    20.1   

   max_order  
0         16  
1         41  
2          9  
3         20  
4         20

In [11]:
df4['loyalty_flag'] = df4.get('loyalty_flag', pd.Series(dtype='object'))
#change column dtype to conduct creation of flag column

In [12]:
df4.loc[df4['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
#assigning first rule of flag column

In [13]:
df4.loc[(df4['max_order'] <= 40) & (df4['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
#assigning second rule of flag column

In [14]:
df4.loc[df4['max_order'] <= 10, 'loyalty_flag'] = 'New customer'
#assigning tihrd rule of flag column

In [16]:
df4['loyalty_flag'].value_counts(dropna = False)
#command to display counts of flag column with applied rules

loyalty_flag
Regular customer    4039742
Loyal customer      2315834
New customer        1752977
Name: count, dtype: int64

In [17]:
df4.head(15)
#confirm changes in loyalty flag column

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0    2059432    80013            11            1                 16   
1      40246    80029            18            0                  9   
2     648191    80154             4            4                 17   
3    1509851    80263            13            5                  1   
4    1500945    80263            15            0                 11   
5    3177085    80263            18            0                 18   
6    3016839    80263            19            6                 18   
7    1327416    80263            20            0                 13   
8    3260064    80309             1            0                  8   
9    1875410    80309             4            3                 19   
10   1644613    80309             7            2                 11   
11   2469334    80332            13            3                 11   
12   1750242    80425            11            3                  5   
13   2736134    80457             3            5                 14   
14   1908214    80457             5            6                 16   

    days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                     14.0             False     26842.0                3.0   
1                     11.0             False     26842.0                6.0   
2                     19.0             False     26842.0                7.0   
3                     20.0             False     26842.0                3.0   
4                      4.0             False     26842.0                2.0   
5                      8.0             False     26842.0                8.0   
6                      6.0             False     26842.0                6.0   
7                     15.0             False     26842.0                7.0   
8                      NaN              True     26842.0                1.0   
9                     22.0             False     26842.0                1.0   
10                    11.0             False     26842.0                1.0   
11                     2.0             False     26842.0                5.0   
12                    15.0             False     26842.0               12.0   
13                    12.0             False     26842.0                5.0   
14                    16.0             False     26842.0                3.0   

    reordered                 product_name  aisle_id  department_id  prices  \
0         1.0  Boneless Beef Sirloin Steak       122             12    20.1   
1         0.0  Boneless Beef Sirloin Steak       122             12    20.1   
2         0.0  Boneless Beef Sirloin Steak       122             12    20.1   
3         0.0  Boneless Beef Sirloin Steak       122             12    20.1   
4         1.0  Boneless Beef Sirloin Steak       122             12    20.1   
5         1.0  Boneless Beef Sirloin Steak       122             12    20.1   
6         1.0  Boneless Beef Sirloin Steak       122             12    20.1   
7         1.0  Boneless Beef Sirloin Steak       122             12    20.1   
8         0.0  Boneless Beef Sirloin Steak       122             12    20.1   
9         1.0  Boneless Beef Sirloin Steak       122             12    20.1   
10        1.0  Boneless Beef Sirloin Steak       122             12    20.1   
11        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
12        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
13        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
14        1.0  Boneless Beef Sirloin Steak       122             12    20.1   

    max_order      loyalty_flag  
0          16  Regular customer  
1          41    Loyal customer  
2           9      New customer  
3          20  Regular customer  
4          20  Regular customer  
5          20  Regular customer  
6          20  Regular customer  
7          20  Regular customer  
8          14  Regular customer

In [18]:
spending_loyalty = df4.groupby('loyalty_flag')['prices'].agg(['mean', 'median', 'std'])
#command to study spending habits of customers

In [19]:
spending_loyalty.head()
#command to study mean median and standard deviation of customer spending habits

mean  median          std
loyalty_flag                                    
Loyal customer    11.093477     6.8   598.620162
New customer      19.175781     7.0  1073.318272
Regular customer  17.825254     6.9  1010.979913

In [20]:
df4['avg_price_per_user'] = df4.groupby('user_id')['prices'].transform('mean')
#calc average price of orders by customer

In [21]:
df4['spending_flag'] = df4.get('spending_flag', pd.Series(dtype='object'))
#command to change dtype of column before using transform function

In [22]:
df4.loc[df4['avg_price_per_user'] < 10, 'spending_flag'] = 'Low spender'
#first rule of flag column

In [24]:
df4.loc[df4['avg_price_per_user'] >= 10, 'spending_flag'] = 'High spender'
#second rule of flag column

In [25]:
df4['spending_flag'].value_counts(dropna = False)
#command to study counts of new flag column

spending_flag
Low spender     7834829
High spender     273724
Name: count, dtype: int64

In [26]:
df4.head(10)
#command to study dataframe shape with new flag column

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   2059432    80013            11            1                 16   
1     40246    80029            18            0                  9   
2    648191    80154             4            4                 17   
3   1509851    80263            13            5                  1   
4   1500945    80263            15            0                 11   
5   3177085    80263            18            0                 18   
6   3016839    80263            19            6                 18   
7   1327416    80263            20            0                 13   
8   3260064    80309             1            0                  8   
9   1875410    80309             4            3                 19   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                    14.0             False     26842.0                3.0   
1                    11.0             False     26842.0                6.0   
2                    19.0             False     26842.0                7.0   
3                    20.0             False     26842.0                3.0   
4                     4.0             False     26842.0                2.0   
5                     8.0             False     26842.0                8.0   
6                     6.0             False     26842.0                6.0   
7                    15.0             False     26842.0                7.0   
8                     NaN              True     26842.0                1.0   
9                    22.0             False     26842.0                1.0   

   reordered                 product_name  aisle_id  department_id  prices  \
0        1.0  Boneless Beef Sirloin Steak       122             12    20.1   
1        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
2        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
3        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
4        1.0  Boneless Beef Sirloin Steak       122             12    20.1   
5        1.0  Boneless Beef Sirloin Steak       122             12    20.1   
6        1.0  Boneless Beef Sirloin Steak       122             12    20.1   
7        1.0  Boneless Beef Sirloin Steak       122             12    20.1   
8        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
9        1.0  Boneless Beef Sirloin Steak       122             12    20.1   

   max_order      loyalty_flag  avg_price_per_user spending_flag  
0         16  Regular customer            6.264000   Low spender  
1         41    Loyal customer            6.920513   Low spender  
2          9      New customer           11.333333  High spender  
3         20  Regular customer            8.607692   Low spender  
4         20  Regular customer            8.607692   Low spender  
5         20  Regular customer            8.607692   Low spender  
6         20  Regular customer            8.607692   Low spender  
7         20  Regular customer            8.607692   Low spender  
8         14  Regular customer           10.252174  High spender  
9         14  Regular customer           10.252174  High spender

In [27]:
df4['user_frequency'] = df4.groupby('user_id')['days_since_prior_order'].transform('median')
#calc median of days since prior order column by user id

In [28]:
df4['median_days_since_prior'] = df4['user_frequency']
#create new column based on user frequency column

In [29]:
df4.loc[df4['median_days_since_prior'] > 20, 'user_frequency'] = 'Non-frequent customer'
#first rule of flag column

C:\Users\boudl\AppData\Local\Temp\ipykernel_10412\4116722164.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'Non-frequent customer' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df4.loc[df4['median_days_since_prior'] > 20, 'user_frequency'] = 'Non-frequent customer'


In [30]:
df4.loc[(df4['median_days_since_prior'] > 10) & (df4['median_days_since_prior'] <= 20), 'user_frequency'] = 'Regular customer'
#second rule of flag column

In [31]:
df4.loc[df4['median_days_since_prior'] <= 10, 'user_frequency'] = 'Frequent customer'
#third rule of flag column

In [32]:
df4['user_frequency'].value_counts(dropna=False)
# Command to display new flag column counts

user_frequency
Frequent customer        5214046
Regular customer         1883122
Non-frequent customer    1008215
NaN                         3170
Name: count, dtype: int64

In [33]:
df4.head()
#command to display new flag column

order_id  user_id  order_number  day_of_week  order_hour_of_day  \
0   2059432    80013            11            1                 16   
1     40246    80029            18            0                  9   
2    648191    80154             4            4                 17   
3   1509851    80263            13            5                  1   
4   1500945    80263            15            0                 11   

   days_since_prior_order  first_order_flag  product_id  add_to_cart_order  \
0                    14.0             False     26842.0                3.0   
1                    11.0             False     26842.0                6.0   
2                    19.0             False     26842.0                7.0   
3                    20.0             False     26842.0                3.0   
4                     4.0             False     26842.0                2.0   

   reordered                 product_name  aisle_id  department_id  prices  \
0        1.0  Boneless Beef Sirloin Steak       122             12    20.1   
1        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
2        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
3        0.0  Boneless Beef Sirloin Steak       122             12    20.1   
4        1.0  Boneless Beef Sirloin Steak       122             12    20.1   

   max_order      loyalty_flag  avg_price_per_user spending_flag  \
0         16  Regular customer            6.264000   Low spender   
1         41    Loyal customer            6.920513   Low spender   
2          9      New customer           11.333333  High spender   
3         20  Regular customer            8.607692   Low spender   
4         20  Regular customer            8.607692   Low spender   

          user_frequency  median_days_since_prior  
0  Non-frequent customer                     25.0  
1      Frequent customer                      8.0  
2       Regular customer                     19.0  
3      Frequent customer                      8.5  
4      Frequent customer                      8.5

In [34]:
path = r'C:\Users\boudl\Instacart Basket Analysis'

In [35]:
df4.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'OPCC' ,'opc_chunk4.csv'))